sdks/java/io/hadoop-file-system/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopFileSystem.java

```diff
@@ -89,23 +89,21 @@
     ImmutableList.Builder<MatchResult> resultsBuilder = ImmutableList.builder();
     for (String spec : specs) {
       try {
-        Set<Metadata> metadata = new HashSet<>();
-
-        FileStatus[] fileStatuses = fileSystem.globStatus(new Path(spec));
-        if (fileStatuses != null) {
-          for (FileStatus fileStatus : fileStatuses) {
-            if (fileStatus.isFile()) {
-              metadata.add(toMetadata(fileStatus));
-            }
-          }
-        }
-
+        final Set<Metadata> metadata = new HashSet<>();
         if (spec.contains("**")) {
+          // recursive glob
           int index = spec.indexOf("**");
           metadata.addAll(
               matchRecursiveGlob(spec.substring(0, index + 1), spec.substring(index + 1)));
+        } else {
+          // normal glob
+          final FileStatus[] fileStatuses = fileSystem.globStatus(new Path(spec));
+          if (fileStatuses != null) {
+            for (FileStatus fileStatus : fileStatuses) {
+              metadata.add(toMetadata(fileStatus));
```

Comment:
Shouldn't the `if (fileStatus.isFile()) {...}` remain here?

Reply:
Maybe I'm confused with what you are trying to do... "HdfsFileSystems.match can match directory" could mean it does and shouldn't or that you want it to be able to match directories.

Can you open a Jira and document the issue and reference it in the PR please?

Reply:
Expected behavior: `HadoopFileSystem.match("hdfs:///tmp/dir)"` returns `MatchResult` with `metadata` about directory (resourceId etc.)

Current behavior `HadoopFileSystem.match("hdfs:///tmp/dir)"` returns `MatchResult` with empty `metadata`